lib/clean-host.js

```diff
@@ -89,30 +76,45 @@ module.exports = function extractHostname(value) {
     url = '' + url;
   }
 
-  var needsTrimming = checkTrimmingNeeded(url);
-  if (needsTrimming) {
-    url = url.trim();
-  }
-
   var needsLowerCase = checkLowerCaseNeeded(url);
   if (needsLowerCase) {
     url = url.toLowerCase();
   }
 
+  var v6Match = url.match(ipv6LikeRE);
+  if (v6Match) {
+    return v6Match[1];
+  }
+
+  var needsTrimming = checkTrimmingNeeded(url);
+  if (needsTrimming) {
+    url = url.trim();
+  }
+
   // Try again after `url` has been transformed to lowercase and trimmed.
   if ((needsLowerCase || needsTrimming) && isValid(url)) {
     return trimTrailingDots(url);
   }
 
   // Proceed with heavier url parsing to extract the hostname.
   if (!hasPrefixRE.test(url)) {
-    url = '//' + url;
+    url = 'https://' + url;
   }
 
-  var parts = parseURL(url);
-
-  if (parts.hostname) {
-    return trimTrailingDots(parts.hostname);
+  try {
+    var parts = new URL(url);
+
+    if (parts.hostname) {
+      // WHATWG URL parses any integer sequence as an IP, whereas the legacy
+      // node URL module would not. Preserve behavior where non-ip-like strings
+      // will not result in valid hostnames.
+      if (ipLikeRE.test(parts.hostname) && !ipLikeRE.test(value)) {
+        return value;
+      }
```

Comment:
WHATWG URL will parse the string `http://42` as the URL `http://0.0.0.42`, which is not how the legacy URL parser worked (it would fail to parse). This just aligns them. A slight difference in this is that when `value` is not a string, we will no longer coerce it to one when returning it. I believe this is closer to the documented API but is technically a breaking change. LMK if you'd like me to restore the test and the string-coercion.